cpp/src/arrow/compute/kernels/codegen_internal.h

```diff
@@ -224,7 +258,9 @@ using enable_if_not_floating_value = enable_if_t<!std::is_floating_point<T>::val
 
 template <typename T, typename R = T>
 using enable_if_decimal_value =
-    enable_if_t<std::is_same<Decimal128, T>::value || std::is_same<Decimal256, T>::value,
+    enable_if_t<std::is_same<Decimal32, T>::value || std::is_same<Decimal64, T>::value ||
+                    std::is_same<Decimal128, T>::value ||
+                    std::is_same<Decimal256, T>::value,
```

Comment:
In theory, because of how the templates work, as long as the value types have the proper methods implemented (operators and methods) then it *should* be safe, with the exception of dealing with the overflows that needed to be managed with the `GetSumType` and so on which we've already caught

Reply:
By safe I mean: all the things someone assumed when they wrote an `enable_if_decimal` still hold. And you're saying you checked all of them with `GetSumType` being the only one where assumptions didn't hold well.

Reply:
It's a potentially breaking change for user code like:

```
template <typename T>
enable_if_decimal_value<T> SetToMagic(T* ptr) {
  constexpr char MAGIC[256] = ...;
  if constexpr (std::is_same_v<T, Decimal128>) {
    memcpy(ptr, &MAGIC, sizeof(Decimal128));
  } else {
    static_assert(std::is_same_v<T, Decimal256>); // error for Decimal32
    memcpy(ptr, &MAGIC, sizeof(MAGIC));
  }
}
```

I think that risk is outweighed by the greater weirdness of not including the new types however

Reply:
This is an internal header, no user code should ever use anything from here.